File: src/report.cpp
#include <ctime>
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <sys/stat.h>

#include <pcap/pcap.h>

#include "report.h"
#include "util.h"

using namespace std;

static constexpr double kbytes = 1024.0;
static constexpr double mbytes = kbytes * kbytes;

Report::Report(const Config* _conf) : conf(_conf) {}

Report::~Report()
{
  if (report_file.is_open()) {
    report_file.close();
  }
}

void Report::start(uint32_t id) noexcept
{
  if (!config_mode_eval(conf)) {
    return;
  }

  start_id = id;
  time_start = std::chrono::steady_clock::now();
}

void Report::process(const size_t bytes) noexcept
{
  if (!config_mode_eval(conf)) {
    return;
  }

  if (bytes > max_bytes) {
    max_bytes = bytes;
  } else if (bytes < min_bytes || min_bytes == 0) {
    min_bytes = bytes;
  }
  sum_bytes += bytes;

  process_cnt++;
}

void Report::skip(const size_t bytes) noexcept
{
  if (!config_mode_eval(conf)) {
    return;
  }

  skip_bytes += bytes;

  skip_cnt++;
}

template <typename T> inline auto PRINT_PRETTY_BYTES(T bytes) -> std::string
{
  double n = static_cast<double>(bytes) / mbytes;
  if (n < 1) {
    n = static_cast<double>(bytes) / kbytes;
    if (n < 1) {
      return std::string(std::to_string(bytes) + "B");
    } else {
      return std::string(std::to_string(n) + "KB");
    }
  }

  return std::string(std::to_string(n) + "MB");
}

auto Report::open_report_file(time_t launch_time) -> bool
{
  std::string filename;

  if (std::filesystem::is_directory("/report")) {
    filename = std::string("/report/") + config_kafka_topic(conf);
  } else {
    filename = config_kafka_topic(conf);
  }

  report_file.open(filename, ios::out | ios::app);

  return report_file.is_open();
}

void Report::end(uint32_t id, time_t launch_time) noexcept
{
  constexpr int arrange_var = 28;
  if (!config_mode_eval(conf)) {
    return;
  }
  if (!open_report_file(launch_time)) {
    return;
  }

  end_id = id;
  time_now = std::chrono::steady_clock::now();
  time_diff = std::chrono::duration_cast<std::chrono::duration<double>>(
      time_now - time_start);

  if (process_cnt) {
    avg_bytes = static_cast<double>(sum_bytes) / process_cnt;
  }

  report_file.precision(2);
  report_file << fixed;
  report_file << "--------------------------------------------------\n";

  time_t current_time =
      std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
  report_file << left << setw(arrange_var)
              << "Time: " << std::ctime(&current_time);

  struct stat st;
  size_t process_bytes = 0;
  switch (config_input_type(conf)) {
  case InputType::Pcap:
  case InputType::Pcapng:
    if (config_input_type(conf) == InputType::Pcap) {
      report_file << left << setw(arrange_var) << "Input(PCAP): ";
    } else {
      report_file << left << setw(arrange_var) << "Input(PCAPNG): ";
    }
    report_file << config_input(conf);
    process_bytes = sum_bytes + sizeof(struct pcap_file_header);
    break;
  case InputType::Log:
    report_file << left << setw(arrange_var)
                << "Input(LOG): " << config_input(conf);
    // add 1 byte newline character per line
    process_bytes = sum_bytes + process_cnt;
    break;
  case InputType::Nic:
    process_bytes = sum_bytes - (sizeof(pcap_pkthdr) * process_cnt);
    report_file << left << setw(arrange_var)
                << "Input(NIC): " << config_input(conf);
    break;
  default:
    break;
  }
  if (config_input_type(conf) != InputType::Nic &&
      stat(config_input(conf), &st) != -1) {
    report_file << "(" << PRINT_PRETTY_BYTES(st.st_size) << ")\n";
  } else {
    report_file << '\n';
  }

  report_file << left << setw(arrange_var)
              << "Datasource ID: " << (int)config_datasource_id(conf) << "\n";
  report_file << left << setw(arrange_var) << "Input ID: " << start_id << " ~ "
              << end_id << "\n";

  switch (config_output_type(conf)) {
  case OutputType::File:
    report_file << left << setw(arrange_var)
                << "Output(FILE): " << config_output(conf);
    if (stat(config_input(conf), &st) != -1) {
      report_file << "(" << PRINT_PRETTY_BYTES(st.st_size) << ")\n";
    } else {
      report_file << "(invalid)\n";
    }
    break;
  case OutputType::Kafka:
    report_file << left << setw(arrange_var)
                << "Output(KAFKA): " << config_kafka_broker(conf) << "("
                << config_kafka_topic(conf) << ")\n";
    break;
  case OutputType::None:
    report_file << "Output(NONE): \n";
    break;
  }

  report_file << left << setw(arrange_var)
              << "Statistics(Min/Max/Avg): " << min_bytes << "/" << max_bytes
              << "/" << avg_bytes << "bytes\n";
  report_file << left << setw(arrange_var) << "Process Count: " << process_cnt
              << "(" << PRINT_PRETTY_BYTES(process_bytes) << ")\n";
  report_file << left << setw(arrange_var) << "Skip Count: " << skip_cnt << "("
              << PRINT_PRETTY_BYTES(skip_bytes) << ")\n";
  report_file << left << setw(arrange_var)
              << "Elapsed Time: " << time_diff.count() << "s\n";
  report_file << left << setw(arrange_var) << "Performance: "
              << PRINT_PRETTY_BYTES(static_cast<double>(process_bytes) /
                                    time_diff.count())
              << "ps\n";

  if (report_file.is_open()) {
    report_file.close();
  }
}

// vim: et:ts=2:sw=2

File: src/report.rs
use crate::Config;
use crate::{
    config_datasource_id, config_input, config_input_type, config_kafka_broker, config_kafka_topic,
    config_mode_eval, config_output, config_output_type,
};
use bytesize::ByteSize;
use chrono::{DateTime, Duration, Utc};
use std::ffi::CStr;
use std::fs::OpenOptions;
use std::io::{self, Write};
use std::path::{Path, PathBuf};

pub struct Report {
    config: *const Config,
    start_id: u32,
    end_id: u32,
    sum_bytes: usize,
    min_bytes: usize,
    max_bytes: usize,
    avg_bytes: f64,
    skip_bytes: usize,
    skip_cnt: usize,
    process_cnt: usize,
    time_start: DateTime<Utc>,
    time_now: DateTime<Utc>,
    time_diff: Duration,
}

impl Report {
    pub fn new(config: *const Config) -> Self {
        Report {
            config,
            start_id: 0,
            end_id: 0,
            sum_bytes: 0,
            min_bytes: 0,
            max_bytes: 0,
            avg_bytes: 0.,
            skip_bytes: 0,
            skip_cnt: 0,
            process_cnt: 0,
            time_start: Utc::now(),
            time_now: Utc::now(),
            time_diff: Duration::zero(),
        }
    }

    pub fn start(&mut self, id: u32) {
        if unsafe { config_mode_eval(self.config) } == 0 {
            return;
        }

        self.start_id = id;
        self.time_start = Utc::now();
    }

    pub fn process(&mut self, bytes: usize) {
        if unsafe { config_mode_eval(self.config) } == 0 {
            return;
        }

        if bytes > self.max_bytes {
            self.max_bytes = bytes;
        } else if bytes < self.min_bytes || self.min_bytes == 0 {
            self.min_bytes = bytes;
        }
        self.sum_bytes += bytes;
        self.process_cnt += 1;
    }

    pub fn skip(&mut self, bytes: usize) {
        if unsafe { config_mode_eval(self.config) } == 0 {
            return;
        }
        self.skip_bytes += bytes;
        self.skip_cnt += 1;
    }

    #[allow(clippy::too_many_lines)]
    pub fn end(&mut self, id: u32) -> io::Result<()> {
        const PCAP_FILE_HEADER_LEN: usize = 24;
        const PCAP_PKTHDR_LEN: usize = 8;
        const ARRANGE_VAR: usize = 28;

        if unsafe { config_mode_eval(self.config) } == 0 {
            return Ok(());
        }

        let report_dir = Path::new("/report");
        let topic = unsafe { CStr::from_ptr(config_kafka_topic(self.config)) }
            .to_str()
            .expect("valid UTF-8");
        let report_path = if report_dir.is_dir() {
            report_dir.join(topic)
        } else {
            PathBuf::from(topic)
        };
        let mut report_file = OpenOptions::new()
            .append(true)
            .create(true)
            .open(report_path)?;

        self.end_id = id;
        self.time_now = Utc::now();
        self.time_diff = self.time_now - self.time_start;

        #[allow(clippy::cast_precision_loss)] // approximation is ok
        if self.process_cnt > 0 {
            self.avg_bytes = self.sum_bytes as f64 / self.process_cnt as f64;
        }

        report_file.write_all(b"--------------------------------------------------\n")?;
        report_file.write_fmt(format_args!(
            "{:width$}{}\n",
            "Time:",
            self.time_now,
            width = ARRANGE_VAR,
        ))?;
        let input_type = unsafe { config_input_type(self.config) };
        let input = unsafe { CStr::from_ptr(config_input(self.config)) }
            .to_str()
            .expect("valid UTF-8");
        let (header, processed_bytes) = match input_type {
            0 => {
                let processed_bytes = (self.sum_bytes + PCAP_FILE_HEADER_LEN) as u64;
                ("Input(PCAP)", processed_bytes)
            }
            1 => {
                let processed_bytes = (self.sum_bytes + PCAP_FILE_HEADER_LEN) as u64;
                ("Input(PCAPNG)", processed_bytes)
            }
            2 => {
                let processed_bytes = (self.sum_bytes + PCAP_PKTHDR_LEN) as u64;
                ("Input(NIC)", processed_bytes)
            }
            3 => {
                // add 1 byte newline character per line
                let processed_bytes = (self.sum_bytes + self.process_cnt) as u64;
                ("Input(LOG)", processed_bytes)
            }
            _ => ("", 0),
        };
        report_file.write_fmt(format_args!(
            "{:width$}{} ({})\n",
            header,
            input,
            ByteSize(processed_bytes).to_string(),
            width = ARRANGE_VAR,
        ))?;
        report_file.write_fmt(format_args!(
            "{:width$}{}\n",
            "Data source ID:",
            u32::from(unsafe { config_datasource_id(self.config) }),
            width = ARRANGE_VAR,
        ))?;
        report_file.write_fmt(format_args!(
            "{:width$}{}-{}\n",
            "Input ID:",
            self.start_id,
            self.end_id,
            width = ARRANGE_VAR,
        ))?;

        let output_type = unsafe { config_output_type(self.config) };
        match output_type {
            0 => {
                report_file.write_all(b"Output(NONE):\n")?;
            }
            1 => {
                let broker = unsafe { CStr::from_ptr(config_kafka_broker(self.config)) }
                    .to_str()
                    .expect("valid UTF-8");
                report_file.write_fmt(format_args!(
                    "{:width$}{} ({})\n",
                    "Output(KAFKA):",
                    broker,
                    topic,
                    width = ARRANGE_VAR,
                ))?;
            }
            2 => {
                let output = unsafe { CStr::from_ptr(config_output(self.config)) }
                    .to_str()
                    .expect("valid UTF-8");
                let size = if let Ok(meta) = Path::new(input).metadata() {
                    ByteSize(meta.len()).to_string()
                } else {
                    "invalid".to_string()
                };
                report_file.write_fmt(format_args!(
                    "{:width$}{} ({})\n",
                    "Output(FILE):",
                    output,
                    size,
                    width = ARRANGE_VAR,
                ))?;
            }
            _ => {}
        }
        report_file.write_fmt(format_args!(
            "{:width$}{}/{}/{:.2} bytes\n",
            "Statistics (Min/Max/Avg):",
            self.min_bytes,
            self.max_bytes,
            self.avg_bytes,
            width = ARRANGE_VAR,
        ))?;
        report_file.write_fmt(format_args!(
            "{:width$}{} ({})\n",
            "Process Count:",
            self.process_cnt,
            ByteSize(processed_bytes).to_string(),
            width = ARRANGE_VAR,
        ))?;
        report_file.write_fmt(format_args!(
            "{:width$}{} ({})\n",
            "Skip Count:",
            self.skip_cnt,
            ByteSize(self.skip_cnt as u64).to_string(),
            width = ARRANGE_VAR,
        ))?;
        #[allow(clippy::cast_precision_loss)] // approximation is okay
        report_file.write_fmt(format_args!(
            "{:width$}{:.2} sec\n",
            "Elapsed Time:",
            self.time_diff.num_milliseconds() as f64 / 1_000.,
            width = ARRANGE_VAR,
        ))?;
        #[allow(clippy::cast_possible_truncation)] // rounded number
        #[allow(clippy::cast_precision_loss)] // approximation is okay
        #[allow(clippy::cast_sign_loss)] // positive number
        report_file.write_fmt(format_args!(
            "{:width$}{}/s\n",
            "Performance:",
            ByteSize(
                (processed_bytes as f64 / (self.time_diff.num_milliseconds() as f64 / 1_000.))
                    .round() as u64
            )
            .to_string(),
            width = ARRANGE_VAR,
        ))?;
        Ok(())
    }
}
